src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1543,6 +1578,39 @@
 
     annotated-query))
 
+;; Top-level parsing
+
+(def experimental-entities
+  #{:event-counts :aggregate-event-counts})
+
+(defn warn-experimental
+  "Show a warning if the endpoint is experimental."
+  [entity]
+  (when (contains? experimental-entities entity)
+    (log/warn (format
+                "The %s entity is experimental and may be altered or removed in the future."
+                (name entity)))))
+
+(defn parse-query-context
+  "Parses a top-level query with a 'from', validates it and returns the entity and remaining-query in
+  a map."
+  [version query warn]
+  (cm/match
+   query
+   ["from" (entity-str :guard #(string? %)) & remaining-query]
+   (let [remaining-query (cm/match
+                          remaining-query
+                          [(q :guard #(vector? %))] q
+                          [] []
+                          :else (throw (IllegalArgumentException. "Your `from` query accepts an optional query only as a second argument. Check your query and try again.")))
+         entity (keyword (utils/underscores->dashes entity-str))]
+     (when warn
```

Comment:
oh, warn-experimental is for experimental entities only? not for things like from?